Fix TemplateProcessor::fixBrokenMacros to handle whitespaces

chapa opened the pull request:
### Description

This is a proposal to fix #590 in a cleaner way than a global regex on the whole document, which might change parts of the document that must not change.

*Edit: see https://github.com/PHPOffice/PHPWord/pull/1971#issuecomment-970481810*

`TemplateProcessor::fixBrokenMacros()` is a method that strip all tags between `${`, the name of the variable and `}` (let's call them the 3 parts). If there is a `<w:t xml:space="preserve">` in those tags it means that there is a trailing space at the end of the content of the tag, because there is not supposed to be any space between the 3 parts, so the whole must be inside of it in order to keep this trailing space.

Examples:
```xml
<w:r><w:t>before ${</w:t></w:r><w:r><w:t xml:space="preserve">variable} </w:t></w:r><w:r><w:t>after</w:t></w:r>
                                                                       |
                                                              the trailing space

<w:r><w:t>before ${</w:t></w:r><w:r><w:t>variable</w:t></w:r><w:r><w:t xml:space="preserve">} </w:t></w:r><w:r><w:t>after</w:t></w:r>
                                                                                             |
                                                                                    the trailing space
```

I think it's safe to move the opening tag just before the variable, but one thing I'm not sure about because I don't know Open XML very well, is whether it's safe to prepend the variable by `</w:t></w:r><w:r><w:t xml:space="preserve">`. I think it's ok if a `<w:t>` is necessarily in a `<w:r>` and there cannot be multiple `<w:t>` in a `<w:r>` (otherwise we could just remove the `</w:r><w:r>` part).

Expected result of the above examples:
```xml
<w:r><w:t>before </w:t></w:r><w:r><w:t xml:space="preserve">${variable} </w:t></w:r><w:r><w:t>after</w:t></w:r>
```

### Checklist:

- [x] I have run `composer run-script check --timeout=0` and no errors were reported
- [x] The new code is covered by unit tests (check build/coverage for coverage report)
- [ ] I have updated the documentation to describe the changes
=> it's a bug fix

chapa committed:
Fix TemplateProcessor::fixBrokenMacros to handle whitespaces

4rthem commented:
Thank you for the fix!
Could it be merged?

chapa commented:
Actually I found a case where my fix isn't working: when there is styling tags between the `$` and the `}` of a single variable. Because all tags are stripped we lose styling information.

After several tests, I came to the conclusion that the best way to fix those broken macros without breaking anything else was to move all existing tags inside the variable right after it, for example:
```xml
<w:r><w:t>${</w:t></w:r><w:r><w:t>variable1</w:t></w:r><w:r><w:t>}</w:t></w:r>
<!--        └─────────1──────────┘         └─────────2──────────┘          -->
```
should be transformed to:
```xml
<w:r><w:t>${variable1}</w:t></w:r><w:r><w:t></w:t></w:r><w:r><w:t></w:t></w:r>
<!--                  └─────────1──────────┘└─────────2──────────┘         -->
```
That way we just reassemble all parts of the macro together.

But we also need to add `xml:space="preserve"` to all `<w:t>` we move, to handle this situation:
```xml
<w:r><w:t>${variable1</w:t></w:r><w:r><w:t>} ${variable2</w:t></w:r><w:r><w:t>}</w:t></w:r>
```
that would be transformed to:
```xml
<w:r><w:t>${variable1}</w:t></w:r><w:r><w:t xml:space="preserve"> ${variable2}</w:t></w:r><w:r><w:t xml:space="preserve"></w:t></w:r>
<!--                                        └──────────────────┘ │                                  └──────────────────┘          -->
<!--                                            we need this     └─ to preserve this space        (this one does not hurt)        -->
```

I updated the fix to handle this

will2877 commented:
Will there be any movement on this PR in the near future?
I'm experiencing isses with missing whitespaces all over my documents.

chapa commented:
Gently ping @troosan

@will2877 until this is merged you can use [`cweagans/composer-patches`](https://packagist.org/packages/cweagans/composer-patches) with the patch at https://github.com/PHPOffice/PHPWord/commit/f955a818014be369a50371da2445dda1c6aaeaf0.patch

chapa committed:
Handle custom macro chars

Update main regexp so that the full macro is matched (only `{{var}` was matched instead of `{{var}}`, which is a problem because we want to move tags after the whole macro)

chapa commented:
I rebased the branch to be up to date with `master`, this needed some adaptations to handle custom macro chars that wasn't overridable at the time the PR was made.

@Progi1984 can you take a look ?

chapa commented:
Hey @Progi1984, I saw some activity on this repo, let me know if you want me to rebase this PR and adapt it to the latest version if needed.

But you can also let me know if you want me to close it, as it seems to be ghosted since I opened it 4 years ago.

chapa commented:
Thanks @adq-talbot, for what it worth I'm using this fix in production since I wrote it and I haven't noticed any problems so far.
